[InstCombine] Fix poison propagation in select of bitwise fold

nikic opened the pull request:
We're replacing the select with the false value here, but it may be more poisonous if m_Not contains poison elements. Fix this by introducing a m_NotForbidPoison matcher and using it here.

Fixes https://github.com/llvm/llvm-project/issues/89500.

nikic committed:
[InstCombine] Fix poison propagation in select of bitwise fold

We're replacing the select with the false value here, but it may
be more poisonous if m_Not contains poison elements. Fix this
by introducing a m_NotForbidPoison matcher and using it here.

Fixes https://github.com/llvm/llvm-project/issues/89500.

goldsteinn commented on llvm/include/llvm/IR/PatternMatch.h:
```diff
@@ -2596,6 +2602,13 @@ m_Not(const ValTy &V) {
+                      true>
+m_NotForbidPoison(const ValTy &V) {
+  return m_c_Xor(m_AllOnesForbidPoison(), V);
+}
```
missing unit tests?

nikic replied:
I prefer not to have unit tests for anything that has lit test coverage.

nikic replied:
(Except for exhaustive unit tests.)

goldsteinn replied:
Err I was confusing, not the `m_NotForbidPoison`, but for the `AllowPoison` template arg in the `cstval_pred_ty` unittests.

Either way though.

goldsteinn commented:
LGTM w/ preference for unit tests for `AllowPoison`

dtcxzyw reviewed (approved):
LGTM.

nikic committed:
Add unit tests for m_NotForbidPoison

This is kind of the replacement for the m_NotForbidUndef matcher
we used to have.
